### PR TITLE
Consider types with ref fields as managed

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/63104

Relates to test plan https://github.com/dotnet/roslyn/issues/59194

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs [hunk 1]
```diff
@@ -185,6 +184,12 @@ private static (bool definitelyManaged, bool hasGenerics) DependsOnDefinitelyMan
                         continue;
                     }
 
+                    if (field.RefKind != RefKind.None)
+                    {
+                        // A ref struct which has a ref field is never considered unmanaged
+                        return (true, hasGenerics);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs:190]: >hasGenerics [](http://example.com/codeflow?start=38&length=11)

Is `hasGenerics` correct at this point, or do we need to look at `field.Type` to determine that?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 2]
```diff
@@ -6345,6 +6345,72 @@ class Program
...
+                // (4,5): error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('StructWithIndirectRefField')
+                //     StructWithIndirectRefField* p = stackalloc StructWithIndirectRefField[10]; // 1, 2
+                Diagnostic(ErrorCode.ERR_ManagedAddr, "StructWithIndirectRefField*").WithArguments("StructWithIndirectRefField").WithLocation(4, 5),
+                // (4,48): error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('StructWithIndirectRefField')
+                //     StructWithIndirectRefField* p = stackalloc StructWithIndirectRefField[10]; // 1, 2
+                Diagnostic(ErrorCode.ERR_ManagedAddr, "StructWithIndirectRefField").WithArguments("StructWithIndirectRefField").WithLocation(4, 48)
+                );
+
+            Assert.True(comp.GetTypeByMetadataName("StructWithIndirectRefField").IsManagedTypeNoUseSiteDiagnostics);
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:6412]: >} [](http://example.com/codeflow?start=8&length=1)

Consider adding a generic `ref struct` test case. Perhaps simply modify this test to:
```csharp
...
ref struct StructWithIndirectRefField
{
    public StructWithRefField<int> Field;
}
ref struct StructWithRefField<T>
{
    public ref T RefField;
}
```

File: src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs (hunk 1 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs:190]: Had the same thought. At a glance this only seems correct if the return type is `(bool definitelyManaged, bool? hasGenerics)`. I haven't looked at the consuming code though. Could be it's correct in the way that you state.

Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs:190]: This is only ever called from `GetManagedKind` and the semantics are correct. Basically if `isDefinitelyManaged == false` then we never look at `hasGenerics`. To reduce confusion I'm going to move this to a local function inside of `GetManagedKind`.

**Comment by jaredpar** [comment]
@cston feedback addressed

**Review by cston — APPROVED** [review]

**Review by jaredpar — APPROVED** [review]

**Pull request merged**